src/OmniSharp.MSBuild/ProjectManager.cs

```diff
@@ -75,6 +92,33 @@ public ProjectManager(ILoggerFactory loggerFactory, IEventEmitter eventEmitter,
             _processLoopTask = Task.Run(() => ProcessLoopAsync(_processLoopCancellation.Token));
 
             _onDirectoryFileChanged = OnDirectoryFileChanged;
+
+            if (_options.OnDemandProjectsLoad)
+            {
+                _workspace.DocumentRequested += OnWorkspaceDocumentRequested;
+            }
+        }
+
+        private void OnWorkspaceDocumentRequested(object sender, DocumentRequestedEventArgs args)
+        {
+            // Search and queue for loading C# projects that are likely to reference the requested file
+            string projectDir = Path.GetDirectoryName(args.DocumentPath);
+            while(projectDir != null && projectDir.StartsWith(_environment.TargetDirectory.TrimEnd(Path.DirectorySeparatorChar)))
+            {
+                List<string> csProjFiles = Directory.EnumerateFiles(projectDir, "*.csproj", SearchOption.TopDirectoryOnly).ToList();
```

Comment:
>List<string> csProjFiles = Directory.EnumerateFiles(projectDir, "*.csproj", SearchOption.TopDirectoryOnly).ToList(); [](start = 16, length = 116)

If you turn this feature on, and someone has a project that contains files outside the subtree of the project file, will they be broken?

Reply:
effectively. do you have a better idea for how to search to "relevant" projects? This approach worked well with codebases I had to deal with so far. Would just adding logging a warning in that case be enough in your opinion?

---
In reply to: [229806732](https://github.com/OmniSharp/omnisharp-roslyn/pull/1322#discussion_r229806732) [](ancestors = 229806732)

Reply:
@rchande : I kept thinking about a better general approach and couldn't come up with one. For C# projects it is very rare when C# source file is located NOT under project's directory. I could think of few customization for the search that could be applied but I do not believe it is worth to try solving the problem without having compelling cases in hand. I would be interested in your thoughts on this.

Reply:
Added OnDemandProjectsLoadSearchStopsAt option to allow controlling the top folder to project searches. By default will stop at a git repo root.

Reply:
@dmgonch I don't understand why `.git` is in the list by default. Isn't `.git` where git stores its internal data? No user files (should) be in it...

Reply:
@rchande : this is just a name of a marker folder. When walking up a tree of dirs, once the one is reached containing one of such marker folders (or files) the search will stop. Indeed we will never search in that folder. Does this clarify things?

Reply:
I believe this is resolved now particularly after adding OnDemandProjectsLoadSearchStopsAt

---
In reply to: [230495784](https://github.com/OmniSharp/omnisharp-roslyn/pull/1322#discussion_r230495784) [](ancestors = 230495784)